lib/mongoid/threaded.rb

```diff
@@ -197,27 +197,37 @@ def client_override=(name)
     # Get the current Mongoid scope.
     #
     # @example Get the scope.
-    #   Threaded.current_scope
+    #   Threaded.current_scope(klass)
     #
     # @return [ Criteria ] The scope.
     #
     # @since 5.0.0
-    def current_scope
-      Thread.current[CURRENT_SCOPE_KEY]
+    def current_scope(klass)
+      if Thread.current[CURRENT_SCOPE_KEY]
+        klass_key = klass
+        Thread.current[CURRENT_SCOPE_KEY].each_key { |key| klass_key = key if key <= klass }
+        Thread.current[CURRENT_SCOPE_KEY][klass_key]
+      end
     end
 
     # Set the current Mongoid scope.
     #
     # @example Set the scope.
-    #   Threaded.current_scope = scope
+    #   Threaded.current_scope(scope, klass)
     #
     # @param [ Criteria ] scope The current scope.
     #
     # @return [ Criteria ] The scope.
     #
     # @since 5.0.0
-    def current_scope=(scope)
-      Thread.current[CURRENT_SCOPE_KEY] = scope
```

Comment:
We have to keep the current_scope= method for backwards compatibility. You can define the new method though and use it internally instead of current_scope=